### ISSUE TITLE
Add to cart more than available stock

### ISSUE THREAD

**Issue by lacherogwu** [woocommerce/woocommerce#29162]
Please provide us with the information requested in this bug report. Without these details, we won't be able to fully evaluate this issue. 
Bug reports lacking detail, or for any other reason than to report a bug, maybe closed without action.

**Prerequisites (mark completed items with an [x]):**
- [x] I have carried out troubleshooting steps and I believe I have found a bug.
- [x] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

**Describe the bug**
**Store settings:**
Add to cart behaviour  -> Add to cart behaviour Redirect to the cart page after successful addition (selected)

When having a product with the "manage stock" selected and adding to the cart more than the available stock it redirects straight to the home page without showing any error.

I also tried to access the "woocommerce_add_to_cart" action and many others like
woocommerce_before_add_to_cart
woocommerce_add_to_cart_validation
woocommerce_cart_redirect_after_error

but none of them are triggered if I'm using URL that is exceeded the total available quantity 
for e.g. http://testings.local/?add-to-cart=10&quantity=9999

**Expected behavior**
Redirect to cart with the maximum amount in stock, or at least in the minimum access to these hooks so we developers can fix it manually by settings validation if it exceeded so reset the cart or something.

**Actual behavior**
If I'm adding to cart more than the available quantity for e.g. http://testings.local/?add-to-cart=10&quantity=9999.
So the cart is stuck and now even if I trying to add just 1 to the cart http://testings.local/?add-to-cart=10&quantity=1 it redirects me to the home page, and until I don't go manually to the /cart page, the issue is not resolved and whenever I will try to add the product to the cart I will be sent to home page

**Steps to reproduce the bug (We need to be able to reproduce the bug in order to fix it.)**
Steps to reproduce the bug:
**Store settings:**
1. Enable Add to cart behaviour Redirect to the cart page after successful addition by going to WooCommerce -> Settings -> Add to cart behaviour
2. Create a product and select "Manage stock" in the "Inventory" tab
3. Set a Stock quantity
4. Go to the website and add product to the cart mote than the stock quantity for e.g. http://testings.local/?add-to-cart=10&quantity=9999
5. You can also try to use "woocommerce_add_to_cart" hook and see that the function is not triggered if you adding more than the stock quantity 

**Screenshots**
https://www.loom.com/share/e8fa168cc46a4bb9b310262d73b4112d

**Isolating the problem (mark completed items with an [x]):**
- [x] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [x] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [x] I can reproduce this bug consistently using the steps above.

**WordPress Environment**
We use the [WooCommerce System Status Report](https://docs.woocommerce.com/document/understanding-the-woocommerce-system-status-report/) to help us evaluate the issue. 
Without this report we won't be able to fully evaluate this issue.
<details>
```
### WordPress Environment ###

WordPress address (URL): http://testings.local
Site address (URL): http://testings.local
WC Version: 5.0.0
REST API Version: ✔ 5.0.0
WC Blocks Version: ✔ 4.0.0
Action Scheduler Version: ✔ 3.1.6
WC Admin Version: ✔ 1.9.0
Log Directory Writable: ✔
WP Version: 5.6.1
WP Multisite: –
WP Memory Limit: 256 MB
WP Debug Mode: –
WP Cron: ✔
Language: en_US
External object cache: –

### Server Environment ###

Server Info: nginx/1.16.0
PHP Version: 7.3.5
PHP Post Max Size: 1,000 MB
PHP Time Limit: 1200
PHP Max Input Vars: 4000
cURL Version: 7.54.0
LibreSSL/2.6.5

SUHOSIN Installed: –
MySQL Version: 8.0.16
Max Upload Size: 300 MB
Default Timezone is UTC: ✔
fsockopen/cURL: ✔
SoapClient: ✔
DOMDocument: ✔
GZip: ✔
Multibyte String: ✔
Remote Post: ✔
Remote Get: ✔

### Database ###

WC Database Version: 5.0.0
WC Database Prefix: wp_
Total Database Size: 3.71MB
Database Data Size: 2.34MB
Database Index Size: 1.37MB
wp_woocommerce_sessions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_api_keys: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_attribute_taxonomies: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_downloadable_product_permissions: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_order_items: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_order_itemmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_tax_rates: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_woocommerce_tax_rate_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zones: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_shipping_zone_locations: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_shipping_zone_methods: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_woocommerce_payment_tokens: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_woocommerce_payment_tokenmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_woocommerce_log: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_actions: Data: 0.02MB + Index: 0.11MB + Engine InnoDB
wp_actionscheduler_claims: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_groups: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_actionscheduler_logs: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_commentmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_comments: Data: 0.02MB + Index: 0.08MB + Engine InnoDB
wp_links: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_options: Data: 1.50MB + Index: 0.03MB + Engine InnoDB
wp_postmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_posts: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_term_relationships: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_term_taxonomy: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_termmeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_terms: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_usermeta: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_users: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_admin_note_actions: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_admin_notes: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_category_lookup: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_customer_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_download_log: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_coupon_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_order_product_lookup: Data: 0.02MB + Index: 0.06MB + Engine InnoDB
wp_wc_order_stats: Data: 0.02MB + Index: 0.05MB + Engine InnoDB
wp_wc_order_tax_lookup: Data: 0.02MB + Index: 0.03MB + Engine InnoDB
wp_wc_product_meta_lookup: Data: 0.02MB + Index: 0.09MB + Engine InnoDB
wp_wc_reserved_stock: Data: 0.02MB + Index: 0.00MB + Engine InnoDB
wp_wc_tax_rate_classes: Data: 0.02MB + Index: 0.02MB + Engine InnoDB
wp_wc_webhooks: Data: 0.02MB + Index: 0.02MB + Engine InnoDB

### Post Type Counts ###

attachment: 1
page: 6
post: 2
product: 1

### Security ###

Secure connection (HTTPS): ❌
					Your store is not using HTTPS. Learn more about HTTPS and SSL Certificates.
Hide errors from visitors: ✔

### Active Plugins (1) ###

WooCommerce: by Automattic – 5.0.0

### Inactive Plugins (0) ###


### Settings ###

API Enabled: –
Force SSL: –
Currency: GBP (£)
Currency Position: left
Thousand Separator: ,
Decimal Separator: .
Number of Decimals: 2
Taxonomies: Product Types: external (external)
grouped (grouped)
simple (simple)
variable (variable)

Taxonomies: Product Visibility: exclude-from-catalog (exclude-from-catalog)
exclude-from-search (exclude-from-search)
featured (featured)
outofstock (outofstock)
rated-1 (rated-1)
rated-2 (rated-2)
rated-3 (rated-3)
rated-4 (rated-4)
rated-5 (rated-5)

Connected to WooCommerce.com: –

### WC Pages ###

Shop base: #6 - /shop/
Cart: #7 - /cart/
Checkout: #8 - /checkout/
My account: #9 - /my-account/
Terms and conditions: ❌ Page not set

### Theme ###

Name: Twenty Twenty-One
Version: 1.1
Author URL: https://wordpress.org/
Child Theme: ❌ – If you are modifying WooCommerce on a parent theme that you did not build personally we recommend using a child theme. See: How to create a child theme
WooCommerce Support: ✔

### Templates ###

Overrides: –

### Action Scheduler ###

Complete: 2
Oldest: 2021-02-17 21:11:01 +0000
Newest: 2021-02-17 21:11:01 +0000
```
</details>

**Comment by zhongruige** [comment]
Hi @lacherogwu,

Thank you for taking the time to report this bug, we really appreciate your help. Also, the video you provided was very helpful! I can reproduce it on my end. This issue appears to only affect the Twenty Twenty-One theme as it behaved as expected on the [Storefront](https://woocommerce.com/storefront/) theme.

1. I checked "Redirect to the cart page after successful addition" under `WooCommerce > Settings > Products`
<img width="641" alt="Screen Shot 2021-02-18 at 3 02 43 PM" src="https://user-images.githubusercontent.com/71906536/108427483-772afa00-71fa-11eb-94e1-c9f435749cf2.png">
2. I set the following for the `woo-belt` product (this is product 14 in my environment):
<img width="594" alt="Screen Shot 2021-02-18 at 3 03 50 PM" src="https://user-images.githubusercontent.com/71906536/108427532-8e69e780-71fa-11eb-93eb-aa450eff72a7.png">

I proceeded to test on the following themes:
* Storefront
* Twenty Twenty-one

For each theme, I followed the direct link, e.g. `http://my-site.com/?add-to-cart=14&quantity=9999` without any code changes to test the behavior.

**Storefront**

In Storefront, the redirect behavior worked as expected and the errors did not stack up in the cart:

<img width="1417" alt="Screen Shot 2021-02-18 at 2 32 08 PM" src="https://user-images.githubusercontent.com/71906536/108428561-fbca4800-71fb-11eb-8ea4-186b1f97d50b.png">

<img width="832" alt="Screen Shot 2021-02-18 at 2 32 25 PM" src="https://user-images.githubusercontent.com/71906536/108428580-ff5dcf00-71fb-11eb-8065-a8f67a6c6e1d.png">

**Twenty Twenty-one**

I was able to reproduce this issue on the Twenty Twenty-one theme. After accessing the link, it would redirect me to the home page. Afterward, when visiting the cart, the error messages would be stacked up:

<img width="1316" alt="Screen Shot 2021-02-18 at 2 43 18 PM" src="https://user-images.githubusercontent.com/71906536/108428989-a3e01100-71fc-11eb-9cce-0cbb6e28a3bc.png">

We won’t be able to include this fix in the upcoming release due to the lower priority of this issue compared to others reported. We’re going to add it to our backlog so we can include it in our planning for one of our future releases.

**Comment by lacherogwu** [comment]
Hey, thanks for the information!

I used in Hello Elementor theme and it has the same issue,
Since I would still like to use Hello Elementor theme because all my
website is build with it and if I change to Storefron theme it's corrupt my
website.
There is any custom code I can apply in order to fix it?

Thanks!

**Comment by EdithAllison** [comment]
I have tested this with Twenty Twenty-one theme and can confirm this happens but only with a direct link (URL) and only if different quantities are added (one error message per quantity I tried to put into basket). 

If the product is added either from the product loop (eg shop page) or product details page with the Woo "add to basket" button, it works as expected. 

@lacherogwu are you using the link direclty, ie without the Woo basket button?

**Comment by pierorocca** [comment]
Claude Code notes:


## Issue #29162 — Analysis

### What the user reports

When using `?add-to-cart=10&quantity=9999` on a non-WooCommerce page (e.g. the homepage) with stock management enabled:

1. The cart stays broken — every subsequent add-to-cart redirects to the **homepage** instead of the cart
2. Error notices **stack up** in the session and only display when the user manually visits `/cart`
3. The `woocommerce_add_to_cart` hook never fires

### The actual code path

```
URL: /?add-to-cart=10&quantity=9999  (on homepage / non-WC page)
     │
     ▼
WC_Form_Handler::add_to_cart_action()   [wp_loaded:20]
     │
     ▼
WC()->cart->add_to_cart($product_id, 9999)
     │
     ├─ stock check fails → throw new Exception("not enough stock")
     │
     ├─ catch (Exception $e) → wc_add_notice($e->getMessage(), 'error')
     │
     └─ return false   ← $was_added_to_cart = false
     
     ▼
if ($was_added_to_cart && 0 === wc_notice_count('error'))  → FALSE
     │
     └─ NO redirect happens. Function returns with no exit.
     
     ▼
WordPress continues loading the current page (the homepage)
```

The error notice is now stored in the session cookie. But the current page (homepage, Twenty Twenty-One) **never calls `woocommerce_output_all_notices()`** because `woocommerce_before_shop_loop`, `woocommerce_before_cart` etc. only fire inside WooCommerce page templates. So the notice sits in the session unprinted.

On the *next* add-to-cart attempt, the stale error notice from the previous request causes `wc_notice_count('error') > 0`, so the redirect to cart is again blocked — the issue from #37164 compounds this one.

### Is the problem outside of core?

**Partially yes, partially no.** There are two distinct problems here:

**Problem 1 — Theme-dependent (outside core):** Storefront works because it calls `woocommerce_output_all_notices` inside its `header.php` on *every* page, not just WooCommerce pages. Twenty Twenty-One and Hello Elementor don't do this, so notices have nowhere to render on non-WC pages. This is a **theme responsibility gap** — WooCommerce only guarantees notice output on its own page templates.

**Problem 2 — Core behaviour (inside core):** When `add_to_cart` fails due to stock, `add_to_cart_action()` does nothing — no redirect at all, not even back to the referring page or the cart. The user is left on the homepage with a broken state. A reasonable core fix would be to redirect to the cart (or referer) even on failure, so the WC notice output hooks can fire and display the error. This is squarely a core responsibility.

`★ Insight ─────────────────────────────────────`
- All `woocommerce_output_all_notices` hook registrations in [`[wc-template-hooks.php:319-330](https://claude.ai/epitaxy/plugins/woocommerce/includes/wc-template-hooks.php)`](plugins/woocommerce/includes/wc-template-hooks.php) are on WooCommerce-specific actions. There is **no fallback for non-WC pages** — this is a deliberate but underdocumented design constraint.
- The stacking of errors is a direct consequence of issue #37164 (stale session notices blocking the redirect guard). These two issues are entangled: fix #37164 and the "stuck" subsequent add-to-cart attempts would recover; fix the missing redirect-on-failure and the homepage dead-end disappears.
- Storefront's approach of calling `woocommerce_output_all_notices` globally in `header.php` is **de facto required** for `?add-to-cart=` links to work correctly from any page — but this is never documented as a theme requirement.
`─────────────────────────────────────────────────`

**Bottom line:** The *notice display* failure is a theme issue (Twenty Twenty-One / Hello Elementor don't output WC notices outside WC pages). But the *redirect-on-failure* gap — leaving the user stranded on the homepage with no feedback — is a genuine core deficiency.

**Comment by pierorocca** [comment]
Hi @lacherogwu - thanks for the report.

I know this has been open for a very long time, and I appreciate your patience. We're currently going through some older WooCommerce issues, identifying what still needs fixing, and tidying up what's already fixed or out of scope.

This appears to still be partially relevant, so I've marked it ready for the engineering team to pick up.
